### PR TITLE
Ability to programatically set environment variables in specmatic

### PR DESCRIPTION
**Pull request by SarveshSharma101**
**What**:
A new function has been added to index.ts to set the values of the variables dynamically inside specmatic.json

**Why**:
To dynamically set the desired values of variables inside specmatic.json , so that custom data to the request-body can be passed in the contract test.

**How**:
The function setEnvVariable, takes 3 parameters as input:
1. name - The name of the variable that needs custom data from userEnd
2. value - The value to define the variable with
3. fileName - The Path to the specmatic.json file in your project

It uses 'fs' module to synchronously write the data to the specmatic.json file.

**Checklist**:

- [ ] Documentation added to the README.md
- [x] Tests
- [x] Typescript definitions updated
- [ ] Ready to be merged

### CODE REVIEW

File: specmatic.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+      "provider": "",
+      "test": [],
+      "stub": []
+    }
+  ],
+  "environments": {
+    "local": {
+      "baseurls": {},
+      "variables": {
+        "key": "updated-value"
```
Comment (sahilaug) [inline specmatic.json:13]: initial-value is a more apt for what this signifies

File: src/index.ts [hunk 2]
```diff
@@ -5,5 +5,6 @@ export {
     startTestServer,
     runContractTests,
     installContracts,
-    installSpecs
+    installSpecs,
+    setEnvVariable
```
Comment (sahilaug) [inline src/index.ts:9]: this is a different name we can use `setSpecmaticEnvironment` instead

File: src/lib/__tests__/index.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import execSh from 'exec-sh';
 import fetch from 'node-fetch';
 import path from 'path';
-import { startStubServer, startTestServer, runContractTests, loadDynamicStub, setExpectations, installContracts, installSpecs } from '../';
+import { setSpecmaticEnvVariable, startStubServer, startTestServer, runContractTests, loadDynamicStub, setExpectations, installContracts, installSpecs } from '../';
```
Comment (sahilaug) [inline src/lib/__tests__/index.ts:4]: I would suggest `setSpecmaticEnvironment`

File: src/lib/index.ts [hunk 4]
```diff
@@ -2,6 +2,17 @@ import fetch from 'node-fetch';
 import path from 'path';
 import execSh from 'exec-sh';
 import { specmaticJarPathLocal } from '../config';
+import fs from 'fs';
+
+
+
+export const setSpecmaticEnvVariable = (name: string, value: string, fileName: string) => {
```
Comment (sahilaug) [inline src/lib/index.ts:9]: fileName should not be a parameter in this method. This is an internal detail that needs to be abstracted away from the consumer projects. This way even if the internal details change (eg. if we end up changing the logic, the consumer projects don't get affected in anyway)

File: src/lib/index.ts [hunk 5]
```diff
@@ -2,6 +2,17 @@ import fetch from 'node-fetch';
...
+  const file = require("../../" + fileName)
```
Comment (sahilaug) [inline src/lib/index.ts:10]: can we try using import here

File: src/lib/index.ts [hunk 6]
```diff
@@ -2,6 +2,17 @@ import fetch from 'node-fetch';
...
+  file.environments.local.variables[name] = value
+  fs.writeFileSync("specmatic.json", JSON.stringify(file, null, 2))
+}
+
+export const setEnvVariable = setSpecmaticEnvVariable;
```
Comment (sahilaug) [inline src/lib/index.ts:15]: I would say lets skip this alias. There are aliases created when an older name is already in use. So that if some consumer project is already using that method, their changes don't break when we change the name of the method. But in our case, because its a new method, no one is using it yet, so we can skip it. and stick with `setSpecmaticEnvironement`

File: src/lib/index.ts (hunk 4 above)
Comment (sahilaug) [inline src/lib/index.ts:9]: instead of just (name, value), let's make it ({ name: value}), basically, an object of multiple name values, so that we can update multiple environment variables together

File: src/lib/__tests__/index.ts [hunk 7]
```diff
@@ -108,3 +108,10 @@ test('setExpectations with a different baseUrl for the stub server', () => {
     body: JSON.stringify(mockStub)
   });
 });
+
+test('setSpecmaticEnvVariable updates the environment variable value in the specmatic.json file', () => {
+  const name = "key"
+  const updatevalue = "updated-value"
+  const fileName = "specmatic.json"
+  expect(setSpecmaticEnvVariable(name, updatevalue, fileName))
```
Comment (sahilaug) [inline src/lib/__tests__/index.ts:116]: What are we expecting here?
A good test would be to add another assertion here, where you read the file and make sure the value has changed from "initial-value" to "updated-value"

File: src/lib/index.ts [hunk 8]
```diff
@@ -2,6 +2,17 @@ import fetch from 'node-fetch';
 import path from 'path';
 import execSh from 'exec-sh';
 import { specmaticJarPathLocal } from '../config';
+import fs from 'fs';
+
```
Comment (sahilaug) [inline src/lib/index.ts:6]: we can get rid of the extra whitespace

File: src/lib/__tests__/index.ts [hunk 9]
```diff
@@ -108,3 +108,10 @@ test('setExpectations with a different baseUrl for the stub server', () => {
...
+});
```
Comment (sahilaug) [inline src/lib/__tests__/index.ts:117]: A couple of more tests I can think of, what if the key is not present in specmatic.json, how does our method handle that

Comment (sahilaug) [reply src/lib/__tests__/index.ts:117]: What if specmatic.json doesnt exist, does out method just break, or does it handle an exception gracefully, we can write a test for that as well

Comment (sahilaug) [reply src/lib/__tests__/index.ts:117]: Similary what happens when file doesnt contain environments key and environments doesn't contain variables key, will our method fail gracefully?
This is how tests can drive better code design

File: src/lib/index.ts (hunk 4 above)
Comment (sahilaug) [reply src/lib/index.ts:9]: Instead of that as a first parameter, you should use the environmentName (In our case we are going to use local) but that should be dynamic

**Comment by sahilaug** [comment]
Also can we add this to REAMDE, in the JS helper section, as an Experimental Section

File: specmatic.json (hunk 1 above)
Comment (SarveshSharma101) [reply specmatic.json]: Added changes

File: src/index.ts (hunk 2 above)
Comment (SarveshSharma101) [reply src/index.ts]: Added changes

File: src/lib/__tests__/index.ts (hunk 3 above)
Comment (SarveshSharma101) [reply src/lib/__tests__/index.ts]: Added changes

File: src/lib/index.ts (hunk 4 above)
Comment (SarveshSharma101) [reply src/lib/index.ts]: Done with the changes, I exported the path for specmatic.json from the config.js and importing it inside index.ts

File: src/lib/index.ts (hunk 5 above)
Comment (SarveshSharma101) [reply src/lib/index.ts]: I imported the path and using path.resolve for better results so that the function can perfectly find the file

File: src/lib/index.ts (hunk 6 above)
Comment (SarveshSharma101) [reply src/lib/index.ts]: Added the required change

File: src/lib/index.ts (hunk 4 above)
Comment (SarveshSharma101) [reply src/lib/index.ts]: Do you mean that we should provide the consumer with ability to update/create a new object inside the environments?
For eg:- if the specmatic.json file only has the baselurl and the user wants to add the variable object to it, so he should be able to add is using this function and also the variable name and its value in a key, value pair right ?

File: src/lib/__tests__/index.ts (hunk 7 above)
Comment (SarveshSharma101) [reply src/lib/__tests__/index.ts]: Refactored the test and expecting the boolean value based on the status if updated or not

File: src/lib/index.ts (hunk 8 above)
Comment (SarveshSharma101) [reply src/lib/index.ts:7]: Removed it

File: src/lib/__tests__/index.ts (hunk 9 above)
Comment (SarveshSharma101) [reply src/lib/__tests__/index.ts:118]: Added try-catch to handle the exception for missing file and if the object is not present inside the specmatic.json

File: src/lib/index.ts [hunk 10]
```diff
@@ -1,7 +1,32 @@
...
+type Environment = Record<string, string>
+export const setSpecmaticEnvironement = (environmentName: string, environment: Environment) => {
+  console.log("Reading the file from -> ", path.resolve(specmatic))
+  let file = null
+  let flag = false
+  try {
+    file = require(path.resolve(specmatic))
+    try {
+      for (var x in environment) {
```
Comment (sahilaug) [inline src/lib/index.ts:16]: Looping over each element in the object means, that many io operations. Instead we can prepare the full contents and write them once

File: src/lib/index.ts [hunk 11]
```diff
@@ -1,7 +1,32 @@
...
+        file.environments.local[environmentName][x] = environment[x]
```
Comment (sahilaug) [inline src/lib/index.ts:17]: We should give the ability to set only the variables, not baseUrls

File: package.json [hunk 12]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "specmatic",
+  "name": "specmatic-Environment",
```
Comment (sahilaug) [inline package.json:2]: Why this change? Is this by mistake?

File: src/lib/__tests__/index.ts [hunk 13]
```diff
@@ -13,98 +13,114 @@ const stubDataPath = './data';
...
+//     body: JSON.stringify(mockStub)
+//   });
+// });
+
+test('setSpecmaticEnvVariable updates the environment variable value in the specmatic.json file', () => {
+  const variableObj1: Record<string, string> = { "key1": "updated-value1", "key2": "updated-value2", "key3": "updated-value3", "key4": "updated-value4" }
+  const variableObj2: Record<string, string> = { "key1": "initial-value1", "key2": "initial-value2", "key3": "initial-value3", "key4": "initial-value4" }
+
+  expect(setSpecmaticEnvironement("variables", variableObj1)).toBe(true)
+  expect(setSpecmaticEnvironement("variables", variableObj2)).toBe(true)
```
Comment (sahilaug) [inline src/lib/__tests__/index.ts:117]: we still need to add assertions after reading the file, if the correct things were actually set

File: src/lib/index.ts (hunk 4 above)
Comment (sahilaug) [reply src/lib/index.ts]: no not even baseUrl, the consumer can just set the environment variables

File: src/lib/__tests__/index.ts (hunk 7 above)
Comment (sahilaug) [reply src/lib/__tests__/index.ts]: How does the boolean ensure that the correct thing was updated or not? I might be setting "updated-value-1", but due to a bug in my method in actual the file is setting "updated-value-2". Hence we should actually read the file and assert on its contents

File: package.json (hunk 12 above)
Comment (SarveshSharma101) [reply package.json:2]: To identify uniquely when i do npm link , and not get confused with the specmatic package in my node_modules. Will remove when the code is final and ready to merge

File: src/lib/index.ts (hunk 11 above)
Comment (SarveshSharma101) [reply src/lib/index.ts:17]: Made the required changes, with the capability to alter/change the desired environment

File: src/lib/index.ts (hunk 10 above)
Comment (SarveshSharma101) [reply src/lib/index.ts:16]: Made the required changes

File: src/lib/__tests__/index.ts (hunk 7 above)
Comment (SarveshSharma101) [reply src/lib/__tests__/index.ts]: Added a new function to check if the required file is updated or not

File: src/lib/index.ts [hunk 14]
```diff
@@ -1,7 +1,38 @@
...
+
+export const setSpecmaticEnvironement = (environmentName: string, environment: Environment) => {
+  console.log("Reading the file from -> ", path.resolve(specmatic))
+  let file = null
+
+  try {
+    file = require(path.resolve(specmatic))
+    for (var x in environment) { file.environments[environmentName].variables[x] = environment[x] }
+    fs.writeFileSync(path.resolve(specmatic), JSON.stringify(file, null, 2))
+    return checkSpecmaticEnvironment(environmentName, environment)
```
Comment (sahilaug) [inline src/lib/index.ts:18]: Why are we checking here. The purpose of the method is to setSpecmaticEnvironment. Checking if it is set properly or not is the concern of the test for this method.

File: src/lib/index.ts [hunk 15]
```diff
@@ -1,7 +1,38 @@
...
+
+type Environment = Record<string, string>
+
+export const setSpecmaticEnvironement = (environmentName: string, environment: Environment) => {
+  console.log("Reading the file from -> ", path.resolve(specmatic))
+  let file = null
+
+  try {
+    file = require(path.resolve(specmatic))
+    for (var x in environment) { file.environments[environmentName].variables[x] = environment[x] }
```
Comment (sahilaug) [inline src/lib/index.ts:16]: 1) use a better variableName, `environmentVariable` 
2) there is no need for {} when there is only one statement inside the block

File: src/lib/index.ts [hunk 16]
```diff
@@ -1,7 +1,38 @@
...
+    fs.writeFileSync(path.resolve(specmatic), JSON.stringify(file, null, 2))
+    return checkSpecmaticEnvironment(environmentName, environment)
+  } catch (e) {
+    if (e instanceof TypeError) { console.log(e.toString()) }
```
Comment (sahilaug) [inline src/lib/index.ts:20]: What is the purpose of this if-else? In any case, we are printing the error right? Is there some other kind of error you are expecting here?

Comment (SarveshSharma101) [reply src/lib/index.ts:20]: Mondule_Not_found, TypeError, but you are right this should be generic .There can be other kinds of error like the datatype of parameter passed is not proper or maybe something else

File: src/lib/index.ts (hunk 15 above)
Comment (SarveshSharma101) [reply src/lib/index.ts:16]: Made the following changes

**Comment by SarveshSharma101** [comment]
> Also can we add this to REAMDE, in the JS helper section and mark it as Experimental

Added the function info to the Readme.md file

File: src/lib/index.ts (hunk 14 above)
Comment (SarveshSharma101) [reply src/lib/index.ts]: It is just the name for a function, in that function I am reading the file again and checking whether the right variables are updated or not based on which that check function returns true or false and this returned boolean value, in turn, gets returned from the main function setSpecmaticEnvironment

Comment (SarveshSharma101) [reply src/lib/index.ts]: Made the changes the checking is being done inside the unit test

**Review by sahilaug — APPROVED** [review]
